node/src/components/block_proposer.rs

```diff
@@ -197,251 +281,118 @@ impl BlockProposer {
-            }
-            Event::Buffer { hash, header } => self.add_deploy(Timestamp::now(), hash, *header),
-            Event::FinalizedProtoBlock { block, mut height } => {
-                let (_, deploys, _) = block.destructure();
-                if height > self.state.next_finalized {
-                    // safe to subtract 1 - height will never be 0 in this branch, because
-                    // next_finalized is at least 0, and height has to be greater
-                    self.state.finalization_queue.insert(height - 1, deploys);
-                } else {
-                    let mut effects = self.handle_finalized_block(effect_builder, height, deploys);
-                    while let Some(deploys) = self.state.finalization_queue.remove(&height) {
-                        height += 1;
-                        effects.extend(self.handle_finalized_block(
-                            effect_builder,
-                            height,
-                            deploys,
-                        ));
-                    }
-                    return effects;
-                }
-            }
-            Event::GetChainspecResult {
-                maybe_deploy_config,
-                chainspec_version,
-                request,
-            } => {
-                let deploy_config = maybe_deploy_config.expect("should return chainspec");
-                // Update chainspec cache.
-                self.chainspecs.insert(chainspec_version, deploy_config);
-                let deploys = self.propose_deploys(
-                    deploy_config,
-                    request.current_instant,
-                    request.past_deploys,
-                );
-                return request.responder.respond(deploys).ignore();
-            }
-        }
-        Effects::new()
-    }
-}
-
-#[derive(Debug, Clone)]
-pub struct BlockProposerMetrics {
-    /// Amount of pending deploys
-    pending_deploys: IntGauge,
-    /// registry Component.
-    registry: Registry,
-}
-
-impl BlockProposerMetrics {
-    pub fn new(registry: Registry) -> Result<Self, prometheus::Error> {
-        let pending_deploys = IntGauge::new("pending_deploy", "amount of pending deploys")?;
-        registry.register(Box::new(pending_deploys.clone()))?;
-        Ok(BlockProposerMetrics {
-            pending_deploys,
-            registry,
-        })
+        deploy_config: DeployConfig,
+        block_timestamp: Timestamp,
+        past_deploys: HashSet<DeployHash>,
+    ) -> HashSet<DeployHash> {
+        // deploys_to_return = all deploys in pending that aren't in finalized blocks or
+        // proposed blocks from the set `past_blocks`
+        self.sets
+            .pending
+            .iter()
+            .filter(|&(hash, deploy)| {
+                self.is_deploy_valid(deploy, block_timestamp, &deploy_config, &past_deploys)
+                    && !past_deploys.contains(hash)
+                    && !self.sets.finalized_deploys.contains_key(hash)
```

Comment:
We'll somehow need to make sure that `block_timestamp` is never older than the most recent timestamp used for pruning.

Reply:
See above, I would assume out of scope? Unless we introduced this bug during the port.

Reply:
No, I agree, but it also needs a ticket. :+1:

Reply:
*cough* @fizyk20 ? :)